### PR TITLE
[dynamo][ddp][ac] Fallback to single bucket when higher order op

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104577
* __->__ #104639
* #104565

This helps unblock an internal model. The real fix requires lot of work, which might question the alternate approach of partitioning AOT graphs instead of Dynamo graphs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

File: test/dynamo/test_activation_checkpointing.py [hunk 1]
```diff
@@ -331,6 +331,117 @@ def fn(x, y, z):
         body_function = getattr(cnt.graphs[0], wrap_node.args[0].name)
         self.assertEqual(op_count(body_function), 2)
 
+    @requires_cuda()
+    def test_ddp(self):
+        import os
+
+        import torch.distributed as dist
+        from torch.nn.parallel import DistributedDataParallel as DDP
+        from torch.testing._internal.common_utils import find_free_port
```
Comment (wconstab) [inline test/dynamo/test_activation_checkpointing.py:340]: was it necessary to import these inside the test due to weird import errors from distributed?

File: test/dynamo/test_activation_checkpointing.py [hunk 2]
```diff
@@ -331,6 +331,117 @@ def fn(x, y, z):
...
+        mod = DDP(mod, bucket_cap_mb=1)
+        x = torch.randn(N, N, requires_grad=True)
+        args = (x,)
+
+        backend = "aot_eager"
+        cnt = CompileCounterWithBackend(backend)
+
+        expected = mod(*args)
+        result = torch.compile(mod, backend=cnt)(*args)
+        dist.destroy_process_group()
```
Comment (wconstab) [inline test/dynamo/test_activation_checkpointing.py:382]: i might prefer you move these tests into test/dynamo/distributed.py and use the infra there that manages processgroup creation/destroy.   You could add a case under `class TestSingleProc(DynamoDistributedSingleProcTestCase):`

File: torch/_dynamo/backends/distributed.py [hunk 3]
```diff
@@ -238,7 +238,21 @@ def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
                         buckets[0].param_ids.append(id(param))
             elif node.op == "get_attr":
                 maybe_param = getattr(gm, node.target)
-                if maybe_param.requires_grad and not self._ignore_parameter(
+                if isinstance(maybe_param, torch.fx.GraphModule):
```
Comment (wconstab) [inline torch/_dynamo/backends/distributed.py:241]: is 'is a graphmodule' the best way to identify a higher-order-op?

File: torch/_dynamo/backends/distributed.py [hunk 4]
```diff
@@ -238,7 +238,21 @@ def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
...
+                    # This indicates presence of a higher order op. For now, we
+                    # have no way to break the higher order op into two buckets.
+                    # Allowing higher order ops in the graph also requires
+                    # changes in the split_module, becuase graph splitter
+                    # currently assumes that all the args of all ops are
+                    # tensors, but in the case of higher order ops, it could be
+                    # a graph module. As a workaround, we are shortcircuiting
+                    log.info(
```
Comment (wconstab) [inline torch/_dynamo/backends/distributed.py:249]: probably warning?

**Review by wconstab — COMMENTED** [review]
Mechanically looks mostly ok, some nits.

From a higher level, a couple questions
1) is it the case that someone actually wants to run with a higher-order-op and DDP, and is ok with the ruined DDP perf?  (have they characterized this perf gap?)
2) based on 1, should we make this behavior automatic (as in this PR), or perhaps just error and tell the user to set `ddp optimizer` config to disabled if they want higher order ops?

**Comment by anijain2305** [comment]
> 1. is it the case that someone actually wants to run with a higher-order-op and DDP, and is ok with the ruined DDP perf?  (have they characterized this perf gap?)

Yes, an internal model has activation checkpointing enabled. This results in HigherOrderOps in the Dynamo graph. For this particular model, its ok to have just one bucket for the entire model. 

> 2. based on 1, should we make this behavior automatic (as in this PR), or perhaps just error and tell the user to set `ddp optimizer` config to disabled if they want higher order ops?

I am ok to both. I have a very mild preference for it to be automatic just for a better UX, although with a LOUD warning.

File: test/dynamo/test_activation_checkpointing.py (hunk 2 above)
Comment (anijain2305) [reply test/dynamo/test_activation_checkpointing.py:382]: Yeah, thats probably better. Will move tests there.

**Comment by wconstab** [comment]
> I am ok to both. I have a very mild preference for it to be automatic just for a better UX, although with a LOUD warning.

My assumption is that most users of DDP+AC would consider it a showstopper to use DDP at all without comms overlaps.  Prototyping work might be the obvious exception, or small training jobs where perf isn't really important.  So i'm not sure if automating the ux is a priority, or we should just prioritize getting it to actually work.

I'm ok with landing to unblock, although, it seems like manually disabling DDPOptimizer is an equally good unblock if its for a particular customer that knows the story already.

**Comment by anijain2305** [comment]
Ok, that sounds good. I will change the logic to fail with an actionable message.

**Comment by anijain2305** [comment]
@wconstab I think I incorporated all your comments. Please take a look when you get a chance.

**Review by wconstab — APPROVED** [review]
ok thanks, lgtm! one other thing you could consider is to pre-emptively file an issue to track "DDP + AC isn't supported" and refer to that issue in the warning message.  That might just help keep any user feedback organized

**Comment by anijain2305** [comment]
> ok thanks, lgtm! one other thing you could consider is to pre-emptively file an issue to track "DDP + AC isn't supported" and refer to that issue in the warning message. That might just help keep any user feedback organized

Neat idea!

**Comment by anijain2305** [comment]
xref - https://github.com/pytorch/pytorch/issues/104674

**Pull request closed without merging**